### ISSUE TITLE
The training results of tood cannot be reproduced on single GPU

### ISSUE THREAD

**Issue by zhaohui-coder** [open-mmlab/mmdetection#8290]
**problem**
When I trained Tood on the coco dataset, I found that there was a big gap between the training results and the data provided by mmdetection. I didn't modify any of the code, and the learning rate was modified using auto-scale-lr. At first I thought it was caused by random seeds, so I set the same random seed with reference to the training logs provided by mmdetection, but the gap still didn't narrow. 

![image](https://user-images.githubusercontent.com/53090200/176988764-1727e817-57b0-460b-ad10-4734c6c7a919.png)

**enviroment**
sys.platform: linux
Python: 3.9.12 (main, Jun  1 2022, 11:38:51) [GCC 7.5.0]
CUDA available: True
GPU 0: NVIDIA GeForce RTX 2080 Ti
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 10.2, V10.2.8
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.11.0
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) oneAPI Math Kernel Library Version 2021.4-Product Build 20210904 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.5.2 (Git Hash a9302535553c73243c632ad3c4c80beec3d19a1e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.2
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_37,code=compute_37
  - CuDNN 7.6.5
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=10.2, CUDNN_VERSION=7.6.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.11.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=OFF, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF, 

TorchVision: 0.12.0
OpenCV: 4.6.0
MMCV: 1.5.3
MMCV Compiler: GCC 7.3
MMCV CUDA Compiler: 10.2
MMDetection: 2.25.0+

**Comment by zhaohui-coder** [comment]
Training command is python tools/train.py configs/tood/tood_r50_fpn_1x_coco.py --seed=1408765599 --auto-scale-lr

**Comment by shinya7y** [comment]
Current implementation of [`auto_scale_lr`](https://github.com/open-mmlab/mmdetection/blob/56e42e72cdf516bebb676e586f408b98f854d84c/mmdet/apis/train.py#L71) does not adjust `warmup_iters`.
It may be proper to set `warmup_iters=500 * 8` manually.

The Linear Scaling Rule does not produce exactly the same results.
Gradient accumulation with `GradientCumulativeOptimizerHook` may be a better choice.
Either way, the functionality of `auto-scale-lr` is incomplete.

**Comment by zhaohui-coder** [comment]
Thanks for your help! I get it!

**Comment by zhaohui-coder** [comment]
> Current implementation of [`auto_scale_lr`](https://github.com/open-mmlab/mmdetection/blob/56e42e72cdf516bebb676e586f408b98f854d84c/mmdet/apis/train.py#L71) does not adjust `warmup_iters`. It may be proper to set `warmup_iters=500 * 8` manually.
> 
> The Linear Scaling Rule does not produce exactly the same results. Gradient accumulation with `GradientCumulativeOptimizerHook` may be a better choice. Either way, the functionality of `auto-scale-lr` is incomplete.

I set `warmup_iters=4000` and `lr=0.00125`, but unfortunately, this doesn't seem to work and the gap still didn't narrow.
![image](https://user-images.githubusercontent.com/53090200/178411031-40cf42f5-bf17-459c-a609-53538976f293.png)
![image](https://user-images.githubusercontent.com/53090200/178411084-c5556442-b8c1-4f07-99e8-06db184830b1.png)

**Comment by chhluo** [comment]
@zhaohui-coder Did you use `GradientCumulativeOptimizerHook` ?

**Comment by zhaohui-coder** [comment]
> @zhaohui-coder Did you use `GradientCumulativeOptimizerHook` ?

yeah，I tried to use it，But the hook degraded the performance of the tood model due to the presence of BN in resnet50
![image](https://user-images.githubusercontent.com/53090200/179388929-68b1bb6f-e891-40be-bef7-a52db7105271.png)
![image](https://user-images.githubusercontent.com/53090200/179389003-3b2179bc-c3a6-49b6-87a3-4b47e011d75f.png)

**Comment by zhaohui-coder** [comment]
> @zhaohui-coder Did you use `GradientCumulativeOptimizerHook` ?

@chhluo  This is a warning that occurs when using ‘GradientCumulativeOptimizerHook’
![image](https://user-images.githubusercontent.com/53090200/179389187-fc456f08-fe4b-4d09-af61-1ba5c03c4a58.png)

**Comment by shinya7y** [comment]
The residual gap (0.006) could be caused by a combination of various factors, e.g., TOOD, ATSS, BatchNorm, more general module in detectors, and randomness.
The following settings may help us find the main factor.
- `atss_r50_fpn_1x_coco.py` with `GradientCumulativeOptimizerHook`
- `retinanet_r50_fpn_1x_coco.py` with `GradientCumulativeOptimizerHook`

**Comment by chhluo** [comment]
1. There is no BatchNorm in the model for `tood/tood_r50_fpn_1x_coco.py` except the backbone. And `norm_eval=True` in bakcbone config means that the running stats(mean and std) are frozen in training. So the BatchNorm layers in backbone do not affect the performance, the warning can be ignored.
2. There is a minor difference between `warmup_iters=4000 and lr=0.00125 with single gpu` and `warmup_iters=500 and lr=0.01 with 8 gpus`.  8 consecutive iters use different learning rates when use the former setting.

**Comment by tureckova** [comment]
@zhaohui-coder I am also unsuccessfully trying to reproduce the tood results.
Could you please provide the final config you tuned up to reproduce the tood results? Thank you very much!

**Comment by liang532** [comment]
> 目前执行不整。手动 设置可能比较合适。[`auto_scale_lr`](https://github.com/open-mmlab/mmdetection/blob/56e42e72cdf516bebb676e586f408b98f854d84c/mmdet/apis/train.py#L71)`warmup_iters``warmup_iters=500 * 8`
> 
> 线性缩放则不会产生完全相同的结果。 坡度累加`GradientCumulativeOptimizerHook`可能是更好的选择。 无论哪种方式，其功能`auto-scale-lr`都是不完整的。

hello, warmup_iters=500 * 8, I did not find papers about it, I want to know the theory about.

**Comment by zhaohui-coder** [comment]
> @zhaohui-coder I am also unsuccessfully trying to reproduce the tood results. Could you please provide the final config you tuned up to reproduce the tood results? Thank you very much!
If you train TOOD on 8 GPUs, it can be reproduced according to the official config, but if you train the model on sigle GPU, there will be a performance gap. I have not solved this problem so far.

**Comment by zhaohui-coder** [comment]
@tureckova If you train TOOD on 8 GPUs, it can be reproduced according to the official config, but if you train the model on sigle GPU, there will be a performance gap. I have not solved this problem so far.

**Comment by zimoqingfeng** [comment]
> 
@zhaohui-coder Could you reproduce the final results with 8 GPUs? I still faced a little performance gap when reproducing it (42.1 vs 42.4)

**Comment by zhaohui-coder** [comment]
@zimoqingfeng I don't have 8 GPUs, but when i reproduced it on 4 GPUs, the mAP is 42.2 on coco valid-set, and the final results are significantly better than using single GPU training(42.0)

**Comment by zimoqingfeng** [comment]
> @zimoqingfeng I don't have 8 GPUs, but when i reproduced it on 4 GPUs, the mAP is 42.2 on coco valid-set, and the final results are significantly better than using single GPU training(42.0)

Same results for me, when training it with 8 GPUs.
